File: src/sde_mcp_server/tools/scans.py
"""Repository scanning tools"""
import json
from typing import Optional

from fastmcp import Context

from ..server import mcp, api_client, init_api_client


@mcp.tool()
async def list_scan_connections(ctx: Context) -> str:
    """List repository scan connections"""
    global api_client
    if api_client is None:
        api_client = init_api_client()
    result = api_client.list_team_onboarding_connections()
    return json.dumps(result, indent=2)


@mcp.tool()
async def scan_repository(ctx: Context, project_id: int, connection_id: int, repository_url: str) -> str:
    """Scan a repository"""
    global api_client
    if api_client is None:
        api_client = init_api_client()
    data = {"project": project_id, "connection": connection_id, "repository_url": repository_url}
    result = api_client.create_team_onboarding_scan(data)
    return json.dumps(result, indent=2)


@mcp.tool()
async def get_scan_status(ctx: Context, scan_id: int) -> str:
    """Get status of a repository scan"""
    global api_client
    if api_client is None:
        api_client = init_api_client()
    result = api_client.get_team_onboarding_scan(scan_id)
    return json.dumps(result, indent=2)


@mcp.tool()
async def list_scans(ctx: Context, project_id: Optional[int] = None) -> str:
    """List repository scans"""
    global api_client
    if api_client is None:
        api_client = init_api_client()
    params = {}
    if project_id is not None:
        params["project"] = project_id
    result = api_client.list_team_onboarding_scans(params)
    return json.dumps(result, indent=2)

File: src/tools/scans.ts
/**
 * Repository scanning tools
 * Translated from: https://github.com/sdelements/sde-mcp/blob/master/src/sde_mcp_server/tools/scans.py
 */

import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import {
  SDElementsClient,
  type SDElementsQueryParams,
} from "../utils/apiClient.js";
import { jsonToolResult } from "./_shared.js";

/**
 * Register all repository scanning tools
 */
export function registerScanTools(
  server: McpServer,
  client: SDElementsClient
): void {
  // List scan connections
  server.registerTool(
    "list_scan_connections",
    {
      title: "List Scan Connections",
      description: "List repository scan connections",
      inputSchema: z.object({}),
    },
    async () => {
      const result = await client.listTeamOnboardingConnections();

      return jsonToolResult(result);
    }
  );

  // Scan repository
  server.registerTool(
    "scan_repository",
    {
      title: "Scan Repository",
      description: "Scan a repository",
      inputSchema: z.object({
        project_id: z.number().describe("ID of the project"),
        connection_id: z.number().describe("ID of the scan connection"),
        repository_url: z.string().describe("URL of the repository to scan"),
      }),
    },
    async ({ project_id, connection_id, repository_url }) => {
      const data = {
        project: project_id,
        connection: connection_id,
        repository_url,
      };

      const result = await client.createTeamOnboardingScan(data);

      return jsonToolResult(result);
    }
  );

  // Get scan status
  server.registerTool(
    "get_scan_status",
    {
      title: "Get Scan Status",
      description: "Get status of a repository scan",
      inputSchema: z.object({
        scan_id: z.number().describe("ID of the scan"),
      }),
    },
    async ({ scan_id }) => {
      const result = await client.getTeamOnboardingScan(scan_id);

      return jsonToolResult(result);
    }
  );

  // List scans
  server.registerTool(
    "list_scans",
    {
      title: "List Scans",
      description: "List repository scans",
      inputSchema: z.object({
        project_id: z.number().optional().describe("Filter by project ID"),
      }),
    },
    async ({ project_id }) => {
      const params: Record<string, unknown> = {};

      if (project_id !== undefined) {
        params.project = project_id;
      }

      const result = await client.listTeamOnboardingScans(
        params as SDElementsQueryParams
      );

      return jsonToolResult(result);
    }
  );
}
